### PR TITLE
Add regex matching for whitelist

### LINKED ISSUE #1: Implement regex matching for LicenseFinder
## Current state

To whitelist licenses for license finder, users can use the command

```
$ license_finder whitelist add "Apache 2.0"
```

However, different projects could list the same license in many ways. For example, the Apache license could be specified as "Apache 2.0", "Apache License (2.0)", "apache, version 2.0". 

This means that multiple whitelist entries have to be added for all the different ways a license can be spelt.

```
$ license_finder whitelist add "Apache License (2.0)"
$ license_finder whitelist add "apache, version 2.0"
```

Furthermore, some projects use multiple licenses. E.g. "MIT or Apache v2 or GPL v3", making it even harder to whitelist all the dependencies that uses the Apache license.
## What we want

We want to be able to whitelist each license with only a single regex whitelist entry in the `dependecy_decisions.yml` file. For example, to whitelist Apache for the above, we only need to whitelist `"/apache/i"`

```
$ license_finder whitelist add "/apache/i"
```
## General acceptance criteria

Please see [CONTRIBUTING.md](https://github.com/idagds/LicenseFinder/blob/master/CONTRIBUTING.md) for the project's contribution guidelines.

Appropriate tests should be written for each additional feature.

In addition to these guidelines, any additional code must not break the CI for any build configuration.

As we do not own the project, the task is considered done once your pull request has been merged into [idagds/LicenseFinder](https://github.com/idagds/LicenseFinder). To benefit the rest of the open source community, we strongly encourage you to also submit a pull request to [pivotal/LicenseFinder](https://github.com/pivotal/LicenseFinder).

### PR DESCRIPTION
**Pull request by linhany**
Fixes #1

### CODE REVIEW

File: lib/license_finder/decisions.rb [hunk 1]
```diff
@@ -31,7 +31,17 @@ def approved?(name, version=nil)
     end
 
     def whitelisted?(lic)
-      @whitelisted.include?(lic)
+      @whitelisted.include?(lic) || whitelist_regex_match?(lic)
+    end
+
+    def whitelist_regex_match?(lic)
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb:37]:
``` ruby
def whitelist_regex_match?(lic)
  @whitelisted.any? { |l| lic.name.to_s =~ RegexParser.parse(l.name) }
end
```

File: lib/license_finder/decisions.rb [hunk 2]
```diff
@@ -210,3 +220,21 @@ def write!(value, file)
...
+module LicenseFinder
+  class RegexParser
+    def self.parse(s)
+      optmap = {
+        "i" => Regexp::IGNORECASE,
+        "m" => Regexp::MULTILINE,
+        "x" => Regexp::EXTENDED,
+      }
+
+      match = s.match(/\/(.*)\/(.*)/) or return
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb]: Can you explain what you are trying to achieve here?

File: lib/license_finder/decisions.rb [hunk 3]
```diff
@@ -210,3 +220,21 @@ def write!(value, file)
...
+      pat = match.captures[0]
+      opt_str = match.captures[1]
+      opts = opt_str.split(//).map { |c| optmap[c] }.reduce { |x, n| x | n }
+      Regexp.new(pat, opts)
+    end
+  end
+end
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb]: Please add a trailing new line.

File: lib/license_finder/decisions.rb [hunk 4]
```diff
@@ -210,3 +220,21 @@ def write!(value, file)
     end
   end
 end
+
+module LicenseFinder
+  class RegexParser
+    def self.parse(s)
+      optmap = {
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb]: `optmap` can be a constant.

File: lib/license_finder/decisions.rb [hunk 5]
```diff
@@ -210,3 +220,21 @@ def write!(value, file)
...
+        "i" => Regexp::IGNORECASE,
+        "m" => Regexp::MULTILINE,
+        "x" => Regexp::EXTENDED,
+      }
+
+      match = s.match(/\/(.*)\/(.*)/) or return
+      pat = match.captures[0]
+      opt_str = match.captures[1]
+      opts = opt_str.split(//).map { |c| optmap[c] }.reduce { |x, n| x | n }
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb]:
``` ruby
opts = opt_str.each_char.map{ |c| optmap[c] }.reduce { |opt, all_opt| opt | all_opt }
```

**Review by lawliet89 — COMMENTED** [review]
Seems to be within the acceptance criteria. Please add tests.

**Comment by linhany** [comment]
Thanks for the suggestions. I have updated the code and added a few tests, let me know if there's anything else!

File: lib/license_finder/decisions.rb (hunk 3 above)
Comment (linhany) [reply lib/license_finder/decisions.rb]: Oops I misunderstood what you meant, please see the latest commit for fix.

File: lib/license_finder/decisions.rb [hunk 6]
```diff
@@ -210,3 +214,21 @@ def write!(value, file)
...
+module LicenseFinder
+  class RegexParser
+    OPTMAP = {
+        "i" => Regexp::IGNORECASE,
+        "m" => Regexp::MULTILINE,
+        "x" => Regexp::EXTENDED,
+      }
+
+    def self.parse(s)
+      match = s.match(/\/(.*)\/(.*)/) or return
```
Comment (lawliet89) [inline lib/license_finder/decisions.rb]: The regex can be improved to `/^\/(.+)\/(.*)$/`

The changes are:
- The whole string must be a regex to work. So test strings like `sometfoobar/apache/i` won't work.
- We don't want to match `//` so I've changed the first capture group to be `(.+)` instead.

Please add tests for both changes.

**Review by lawliet89 — CHANGES_REQUESTED** [review]
Good to go except for the regex changes.

Don't worry about the test failures for `Ruby < 2.2.2` for now.  We'll go look into them.

**Comment by linhany** [comment]
Added the regex change and tests to ensure it works as expected.

**Review by lawliet89 — APPROVED** [review]

**Comment by lawliet89** [comment]
Thanks for the work. I am not merging this in yet, while I await some changes from upstream.

Please login to [govBuy](https://buy.gds-gov.tech/) and mark your auction as done. We will email you separately for payment details.

**Comment by linhany** [comment]
Ok, thanks so much for the help! :)

**Pull request closed without merging**